### ISSUE TITLE
AdditionTextFile.GetText has a single lock

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#41301]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/849582/vs2019-csc-causes-high-cpu-load-100.html)._

---
Hi,
I have three open solutions in Visual Studio 16.4.0. Suddenly the computer becomes very slow and if I check the taskmanager, I can see 3x csc and 1x VBCSCompiler tasks. Every task > 20 %. The load of Visual Studio is about 94%. I'm not in a debugging session, I was just reading an article in the browser as the mouse and keyboard become very slow.

I'm not able to record something. :-(

---
#### Mike Küster on 12/9/2019, 07:31 AM: 

<p>Hi,</p><p>I found steps to reproduce: </p><ul><li>Clone this repo: https://github.com/SixLabors/ImageSharp</li><li>Open the solution</li><li>Start Rebuild</li></ul><p>Result:</p><p>2x csc and 1x VBCSCompiler Task with about 95%</p><p>Output:</p><p>1&gt;------ Rebuild All started: Project: ImageSharp, Configuration: Debug Any CPU ------<br>1&gt;C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.CSharp.CurrentVersion.targets(135,9): warning MSB3884: Could not find rule set file "C:\source\Spielwiese\ImageSharp\src\..\shared-infrastructure\SixLabors.ruleset".<br>1&gt;C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.CSharp.CurrentVersion.targets(135,9): warning MSB3884: Could not find rule set file "C:\source\Spielwiese\ImageSharp\src\..\shared-infrastructure\SixLabors.ruleset".<br>1&gt;C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.CSharp.CurrentVersion.targets(135,9): warning MSB3884: Could not find rule set file "C:\source\Spielwiese\ImageSharp\src\..\shared-infrastructure\SixLabors.ruleset".<br>1&gt;C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.CSharp.CurrentVersion.targets(135,9): warning MSB3884: Could not find rule set file "C:\source\Spielwiese\ImageSharp\src\..\shared-infrastructure\SixLabors.ruleset".</p><p>If I cancel the build process, the VBCSCompiler keeps running, even if I try to close Visual Studio. I have to cancel the task with the TaskManager ...</p>

#### Roi Chen [MSFT] on 12/9/2019, 06:46 PM: 

<p>Thank you for taking the time to log this issue!<br>
I’ve tried to reproduce and investigate using the description, and attachments already provided. Unfortunately those aren’t enough and more information is needed in order to investigate it further.<br>
The easiest way to provide all the information is to use the Visual Studio Feedback Tool. This will ensure that we collect the needed information for you without worrying about what to provide (recording, dump file or ETL trace).<br>
Since this issue is now marked as Need More Info, that workflow is enabled in the Feedback Tool:<br>
• Open Visual Studio Feedback tool.<br>
• Click the banner letting you know that you have problems requesting your attention.<br>
• Click this problem from the list<br>
• Click “View their request and respond” from the problem details banner<br>
• Add a comment, in the Attachments/Record: click Start Recording<br>
• When the Steps Recorder tool appears, perform the steps that reproduce the problem.<br>
• When you’re done, choose the Stop Record button.<br>
• Wait a few minutes for Visual Studio to collect and package the information that you recorded.<br>
• Submit. You will be able to see the comment on Developer Community. For security reasons, your files come directly to us and don’t appear on Developer Community.<br>
For the full instructions, please see: <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/how-to-report-a-problem-with-visual-studio-2017?view=vs-2017#when-further-information-is-needed-need-more-info">https://docs.microsoft.com/en-us/visualstudio/ide/how-to-report-a-problem-with-visual-studio-2017?view=vs-2017#when-further-information-is-needed-need-more-info</a> . For information about what data is collected, see <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/developer-community-privacy?view=vs-2017#data-we-collect">https://docs.microsoft.com/en-us/visualstudio/ide/developer-community-privacy?view=vs-2017#data-we-collect</a><br>
We look forward to hearing from you!</p>

#### Mike Küster on 12/10/2019, 00:49 AM: 

Hi, here is the requested recording ....<br><br>First I cloned https://github.com/SixLabors/ImageSharp.git into a new folder.<br>
#### Mike Küster on 12/10/2019, 01:12 AM: 

<p>If I disable the Code Analysis : Run on Build in the ImageSharp properties, I can compile the code (with errors) ...</p>

#### Visual Studio Feedback System on 12/16/2019, 01:06 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Visual Studio Feedback System on 1/16/2020, 05:44 PM: 

I have detected that for the last 35 days, this issue didn't have much product team activity and a very small amount of new votes or comments. Based on this, its severity, and affected area, it’s my experience that this issue is very unlikely to be fixed.

---
### Original Solutions
(no solutions)

**Comment by agocke** [comment]
Looks like the root cause here is the StyleCop analyzers calling AdditionalTextFile.GetText a ton, and then everyone gets stuck in a lock

cc @sharwell

**Comment by sharwell** [comment]
@agocke The method should not be taking a lock if `_text` is already initialized.

**Comment by agocke** [comment]
Double-checked locking is indeed one way to fix this

**Comment by sharwell** [comment]
StyleCop Analyzers is planning to support .editorconfig as a way to configure the analyzer behavior. Users who switch to this would not need to include stylecop.json, and wouldn't hit the `GetText` issue. Double-checked locking would resolve issues for remaining users who prefer the json format.

**Comment by dnchattan** [comment]
I have been plagued by this issue for several months and followed all the suggestions here, as well as in similar threads. Unfortunately in my case it was not being caused by an analyzer, but rather appears to be GC related in my case.

This problem regularly reproduces in my project (https://github.com/raid-toolkit/raid-toolkit-sdk), simply running an `msbuild` on the SDK.sln will spawn a `VBCSCompiler.exe` process which then lives for some time (default is 600s in the shipped `.exe.config`). I can reproduce this simply by triggering the one build, and can even close the terminal/VS/etc-- at some point shortly after the CPU pegs to 100% on all cores and the entire system becomes unresponsive. After a minute or two it will go back down and the system will start responding again.

I grabbed a couple memory dumps of the process while this happened and every one that occurred while the CPU was pegged had nearly every managed thread blocked for GC.

Without any other ideas I updated `VBCSCompiler.exe.config` to include the following `GCHeapCount`/`GCHeapAffinitizeMask` elements under `<runtime>` to prevent the process from pegging all cores:
```xml
    <GCHeapCount enabled="10"/>
    <GCHeapAffinitizeMask enabled="15"/>
```

After making this change, not only does it not bring my system down, but it actually no longer pegs the CPU at all on my system even during large builds.

Obviously, this is not an ideal solution, and not the root of my flavor of this issue. It may be netcore GC related, or somehow exacerbated by how this tool affects the managed heap when it runs- but hopefully this will either help someone else or help you narrow down the root cause and a better fix :).
